examples/particle-swarm-optimizer.dx

```diff
@@ -0,0 +1,96 @@
+' # Particle Swarm Optimizer
+
+' ## Fitness function
+
+rosenbrock:: Real -> Real -> Real
+rosenbrock x y = sq (1.0 - x) + 10000.0*sq (y - x*x) 
+
+
+' We write one that uses vector for input
+
+rosenbrock2:: (2=>Real) -> Real
+rosenbrock2 xs =
+  x = xs.(asidx 0)
+  y = xs.(asidx 1)
+  rosenbrock x y
+
+
+' Min should be at 1.0, 1.0
+
+:p rosenbrock 1.0 1.000
+
+:p rosenbrock2 [1.0, 1.000]
+
+:p rosenbrock 1.0 1.02
+
+:p rosenbrock2 [1.0, 1.02]
+
+
+
+' ## Helper functions
+
+' make a random vector unifornly distrubuted between lb and ub
+
+randBounded:: Key -> (d=>Real)->(d=>Real)->(d=>Real)
+randBounded key lb ub =
+    for i. lb.i + ((rand $ ixkey key i) * (ub.i - lb.i))
+
+randBounded (newKey 4) [1.0,  -2.0] [-1.0,  2.0]
+
+' minby and minimumby
+' to find the smallest values
+
+minby:: A a::Data. (a->Real)->a->a->a
+minby f x y = [x,y].(asidx $ b2i $ (f x) > (f y))
+
+minby sq -1.0 0.5
+minby fst (0.7, 1000.0) (3.0, 12.0)
+
+
+minimumby:: A p q::Data. (q->Real)->(p=>q)->q
+minimumby f xs = fold xs.(asidx 0) (for i. minby f xs.i)
+
+minimumby sq [-1.0, -0.1, 2.0, 3.0]
+
+
```

Comment:
Should these `minby` and `minimumby` move to the preable.dx?
They seem useful

and also would add `min = minby id` and similar for `minimum`

Reply:
Yes, good idea. As you've noticed, we can now (since 63259fb11690a) implement them more cleanly using `select` (once we fix #34!)